Zend/zend_observer.h

```diff
@@ -0,0 +1,111 @@
+   +----------------------------------------------------------------------+
+   | Authors: Levi Morrison <levim@php.net>                               |
+   |          Sammy Kaye Powers <sammyk@php.net>                          |
+   +----------------------------------------------------------------------+
+*/
+
+#ifndef ZEND_OBSERVER_H
+#define ZEND_OBSERVER_H
+
+#include "zend.h"
+#include "zend_compile.h"
+
+BEGIN_EXTERN_C()
+
+extern ZEND_API int zend_observer_fcall_op_array_extension;
+
+#define ZEND_OBSERVER_ENABLED (zend_observer_fcall_op_array_extension != -1)
+
+#define ZEND_OBSERVER_HANDLERS(op_array) \
+	ZEND_OP_ARRAY_EXTENSION(op_array, zend_observer_fcall_op_array_extension)
+
+#define ZEND_OBSERVER_NOT_OBSERVED ((void *) 2)
+
+#define ZEND_OBSERVABLE_FN(fn_flags) \
+	(ZEND_OBSERVER_ENABLED && \
+		!(fn_flags & (ZEND_ACC_CALL_VIA_TRAMPOLINE | ZEND_ACC_FAKE_CLOSURE)))
+
+struct zend_observer_fcall {
+	void (*begin)(zend_execute_data *execute_data);
+	void (*end)(zend_execute_data *execute_data, zval *retval);
+};
+typedef struct zend_observer_fcall zend_observer_fcall;
+
+struct zend_observer_fcall_cache {
+	// points after the last handler
+	zend_observer_fcall *end;
+	// a variadic array using "struct hack"
+	zend_observer_fcall handlers[1];
+};
+typedef struct zend_observer_fcall_cache zend_observer_fcall_cache;
+
+/* If the fn should not be observed then return {NULL, NULL} */
+typedef zend_observer_fcall(*zend_observer_fcall_init)(zend_function *func);
+
+// Call during minit/startup ONLY
+ZEND_API void zend_observer_fcall_register(zend_observer_fcall_init init);
+
+ZEND_API void zend_observer_startup(void); // Called by engine before MINITs
+ZEND_API void zend_observer_activate(void);
+ZEND_API void zend_observer_deactivate(void);
+ZEND_API void zend_observer_shutdown(void);
+
+ZEND_API void zend_observer_fcall_install(zend_function *function);
+
+ZEND_API void zend_observe_fcall_begin(
+	zend_observer_fcall_cache *cache,
+	zend_execute_data *execute_data);
+
+ZEND_API void zend_observe_fcall_end(
+	zend_observer_fcall_cache *cache,
+	zend_execute_data *execute_data,
+	zval *return_value);
+
+ZEND_API void zend_observer_fcall_call_end_helper(
+	zend_execute_data *execute_data,
+	zval *return_value);
+
+ZEND_API zend_always_inline void zend_observer_maybe_fcall_call_begin(
+	zend_execute_data *execute_data)
+{
+	zend_op_array *op_array = (zend_op_array *)execute_data->func;
+	if (ZEND_OBSERVABLE_FN(op_array->fn_flags) &&
+		!(op_array->fn_flags & ZEND_ACC_GENERATOR)) {
+		void *observer_handlers = ZEND_OBSERVER_HANDLERS(&EX(func)->op_array);
+		if (!observer_handlers) {
+			zend_observer_fcall_install((zend_function *)&EX(func)->op_array);
+			observer_handlers = ZEND_OBSERVER_HANDLERS(&EX(func)->op_array);
+		}
+		ZEND_ASSERT(observer_handlers);
+		if (observer_handlers != ZEND_OBSERVER_NOT_OBSERVED) {
+			zend_observe_fcall_begin(observer_handlers, execute_data);
+		}
```

Comment:
Should the body here get extracted into a helper like I did for `zend_observer_maybe_fcall_call_end` to reduce code size when inlining? Or should the reverse be done and `zend_observer_fcall_call_end_helper` inlined now that these are mostly called only in the `SPEC(OBSERVER)` functions?

Reply:
I don't know what would be better so I don't have any strong opinions on this one.